### ISSUE TITLE
tcpassembly reorder bug

### ISSUE THREAD

**Issue by david415** [google/gopacket#81]
Dear Graeme Connell,

There are a few problems with the TCP "reordering" of tcpassembly:

- If the Assembler receives duplicate out of order packets then it will emit an extra Reassembly struct into the stream. The extra Reassembly struct will contain an empty Bytes payload. This is not desired behavior I think. 

- The Assembler does not ignore coalesce injection packets; they corrupt the reassembly stream.

- The Assembler does not emulate the reordering properties of Selective Acknowledgement. That is, if two out of order packet (with sequence numbers that come after the TCP FSM's "next Sequence") with the same sequence number are received... the second packet's payload is coalesced into the stream whereas in implementations of Selective Acknowledgement the first packet wins. (if my understanding of Selective Ackowledgement is correct)

https://github.com/david415/gopacket/tree/test-reorder
https://github.com/david415/gopacket/commit/89677d3b4ea713a753d343e06b060a24a4c01a89
AND
https://github.com/david415/gopacket/commit/ca61db8ba85cb6bbb1c1d742fb08f9932df48ff6

I've already fixed these problems in Honeybadger... here:
https://github.com/david415/HoneyBadger/blob/51-fix-retrospective-1/ordered_coalesce.go#L315-L325

...that is, I fixed that code snippet and it was inspired by your `addNextFromConn`

**Comment by orebokech** [comment]
Hi David,

Interesting, thanks. Do you have a pcap file to demonstrate these attacks that we could use as a test for the new asssembly package that we're working on @sentryo?

Thanks!

**Comment by gconnell** [comment]
The sending of empty reassembly stuff is actually working as intended, but
the other stuff isn't.

Because the Reassembly struct contains other information besides the actual
bytes (EG: the timestamp a packet was seen), we send it to the stream even
if it contains no new data.

I'm not sure what you mean by "coalesce injection packets"?  Could you
provide any further information on that attack method?

I believe there's currently no code to handle selective ACKs.  This is
definitely an issue we need to address.

If you're interested, there's work starting up on a newer, shinier TCP
assembly module... let us know if that's something you'd be interested in
either working on or providing test cases for.

**Comment by gconnell** [comment]
Ah, wait, I see that you've already commented extensively on the design.
So I assume you know about it :P

**Comment by david415** [comment]
Yes... I generated two pcap files with TCP injection attacks (handshake hijack and sloppy stream injection).. but I also copied the 7 FOX-IT "Quantum Injection" pcap files into this repo:

https://github.com/david415/honeybadger-pcap-files

However, those pcap files do not demonstrate a coalesce injection. What is a coalesce injection? It is a TCP injection attack where the attacker sends TCP packets out of order. Instead of sending the next sequence the attack sends nextSeq+n...

Yes... I'd like to get more involved with the development efforts of this shiny new tcpassembly!

**Comment by gconnell** [comment]
"It is a TCP injection attack where the attacker sends TCP packets out of
order. Instead of sending the next sequence the attack sends nextSeq+n..."

Not sure I follow... how is this different from any other injection?
What's the attack model here... does the attacker have full access to all
the packets (MITM), or is he guessing at sequence numbers?  Are you talking
about doing fuzzy sequence prediction by adding to the current sequence
number?

**Comment by david415** [comment]
TCP injection is used when an attacker has control over a communications channel but not control of a router. This is different than MITM because in the case of MOTS (man on the side) the attacker cannot prevent packets from traversing between Alice and Bob. They can inject "extra" packets into the stream if they discover the sequence number AND win the race.

My rough draft blog post for Tor Project can provide you with even more context than this with it's multitude of reference links to recently leaked NSA documents and writing my other security researchers:

https://github.com/david415/HoneyBadger_docs/blob/hackpad1/source/how-to-badger-the-puppet-masters.rst

You might at least be interested to read about my categorization of all TCP injections attacks.

**Comment by david415** [comment]
A coalesce injection is slightly different than a "ordered injection" in that the actual injection is delayed because the TCP implementation will buffer the segment until it's Sequence become the FSM's next Sequence.

**Comment by david415** [comment]
Modern TCP injection attacks are now understood in the context of the attacker controlling one or more communications channels in the route... and thus the attacker knows the TCP Sequence numbers. The more sophisticated attackers will inject the packets from off-path hosts if ingress filtering doesn't get in the way. The recent China-entity-versus-Github TCP censorship injection attacks is an example of a man-on-the-side attack... and it most definitely required the attack to be reading the TCP Sequence numbers off the wire before sending their injection packets:

http://www.netresec.com/?page=Blog&month=2015-03&post=China%27s-Man-on-the-Side-Attack-on-GitHub

In principal honeybadger detects attacks like this... however I've got a little more work to do before it'll detect injected FIN and RST packets...
